### PR TITLE
LUD-62 ScaleIO puppet model to support discover and inventory of scaleIO cluster and node

### PR DESCRIPTION
**Pull request by sushilrai**
- Included support for fetching all ScaleIO Systems
- SDS
- SDC
- Protection Domains
- Storage Pools
- Volumes
- Fault-sets

### CODE REVIEW

File: bin/discovery.rb [hunk 1]
```diff
@@ -0,0 +1,125 @@
+#!/opt/puppet/bin/ruby
+
+require 'trollop'
+require 'json'
+require 'timeout'
+require 'pathname'
+
+puppet_dir = File.join(Pathname.new(__FILE__).parent.parent,'lib','puppet')
```
Comment (gavin-scott) [inline bin/discovery.rb]: Per our style guidelines please use double quote everywhere and add a space between args.

File: bin/discovery.rb [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+      File.write(@opts[:output], JSON.pretty_generate(JSON.parse(facts)))
+    else
+      results ||= {}
+      scaleio_cache = "/opt/Dell/ASM/cache"
+      Dir.mkdir(scaleio_cache) unless Dir.exists? scaleio_cache
+      file_path = File.join(scaleio_cache, "#{opts[:server]}.json")
+      File.write(file_path, results) unless results.empty?
+    end
+  end
+end
```
Comment (gavin-scott) [inline bin/discovery.rb]: Please add end of line

File: lib/puppet/scaleio/util.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+        require 'asm/device_management'
+        begin
+          ASM::DeviceManagement.parse_device_config(Puppet[:certname])
+        rescue
+          raise Puppet::Error, "Error parsing device config for: #{Puppet[:certname]}"
+        end
+      end
+    end
+  end
+end
```
Comment (gavin-scott) [inline lib/puppet/scaleio/util.rb]: Please add end of line

File: lib/puppet/scaleio/util.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+module Puppet
+  module ScaleIO
+    module Util
+      def self.get_transport
+        require 'asm/device_management'
+        begin
+          ASM::DeviceManagement.parse_device_config(Puppet[:certname])
```
Comment (gavin-scott) [inline lib/puppet/scaleio/util.rb]: We shouldn't use internal code in puppet modules. Can you look at using a transport resource the way that we do with e.g. vcenter and idrac modules? We can then use a custom provider to get the creds as we do for those, see e.g. https://github.com/dell-asm/asm/tree/master/lib/puppet/provider

**Comment by sushilrai** [comment]
@gavin-scott please review

File: lib/puppet/scaleio/util.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+module Puppet
+  module ScaleIO
+    module Util
+      def self.get_transport
+        require "asm/device_management"
+        begin
+          ASM::DeviceManagement.parse_device_config(Puppet[:certname])
```
Comment (gavin-scott) [inline lib/puppet/scaleio/util.rb]: Remaining request from previous review was: "We shouldn't use internal code in puppet modules. Can you look at using a transport resource the way that we do with e.g. vcenter and idrac modules? We can then use a custom provider to get the creds as we do for those, see e.g. https://github.com/dell-asm/asm/tree/master/lib/puppet/provider"

**Comment by sushilrai** [comment]
@gavin-scott Please review

**Pull request merged**